scipy/sparse/coo.py

```diff
@@ -145,8 +145,8 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
                     if len(row) == 0 or len(col) == 0:
                         raise ValueError('cannot infer dimensions from zero '
                                          'sized index arrays')
-                    M = np.max(row) + 1
-                    N = np.max(col) + 1
+                    M = int(np.max(row) + 1)
```

Comment:
Hmmm, that's quite problematic! I guess the only solution would be to use some outside help from Cython.

Although, what I was trying to accomplish was not to be able to have rows/column numbers larger than what `int` can hold, but rather to be able to initialize `coo_matrix` from arrays with type `size_t` without specifying shape, which currently fails as `np.max(row) + 1` returns a float.

Reply:
Probably should be `operator.index(np.max(row)) + 1`

Reply:
A quick Cython fix seems to be able to return an integer bigger than 2^53:
```python
cpdef size_t max_plus_one(arr):
    cdef size_t max_arr = arr.max()
    return max_arr + 1
```
```python
max_plus_one(np.array([2**60]).astype(ctypes.c_size_t))
```
Returns:
```1152921504606846977```
(Which is the correct result)

Also:
```python
type(max_plus_one(np.array([2**60]).astype(ctypes.c_size_t)))
```
```int```

Reply:
Building a `coo_matrix` with that output also seems to work fine:
```python
from scipy.sparse import coo_matrix
coo_matrix((np.array([1.0]), (np.array([1]), np.array([2**60]).astype(ctypes.c_size_t))),
   shape=(2, max_plus_one(np.array([2**60]).astype(ctypes.c_size_t))))
```
```
<2x1152921504606846977 sparse matrix of type '<class 'numpy.float64'>'
	with 1 stored elements in COOrdinate format>
```

Reply:
At a minimum, you just want to avoid the conversion to floating point.  So something like

    M = np.max(row)
    M += type(M)(1)

That avoids the conversion by incrementing `maxrow` with the value 1 represented in the same data type as `np.max(row)`. (One more very extreme case that this doesn't fix is if the maximum row is `2**64-1`.  By adding 1, the integer overflows to 0.)  Another hackish technique is something like

    M = (np.max(row, keepdims=True) + 1)[0]

That works because the cast to floating point is done for numpy *scalars*, but not for numpy *arrays*.  `np.max(row, keepdims=True)` is an array with shape (1,), so adding 1 to it preserves the data type of the array.

Don't be in a rush to make one of these changes.  @pv has a better view of the big picture for sparse array indexing, and could probably suggest a better change.

Reply:
It might be obvious, but I started writing my previous comment before the comment from @pv appeared, and didn't see the followup comments until I submitted it.

Reply:
@pv wrote
> ...since the aim is to have a Python integer in the end.

That makes sense.  So another possibility (that doesn't require import `operator`) is

    M = np.max(row).item() + 1

Reply:
@pv, does `operator.index(np.max(row))` have an advantage over just `int(np.max(row))`?

Reply:
Yes, it will fail for float64 etc.

Reply:
Ah, right. Thanks.